support/src/main/java/com/yubico/yubikit/support/DeviceUtil.java

```diff
@@ -0,0 +1,528 @@
+                false);
+    }
+
+    static DeviceInfo readInfoFido(FidoConnection connection, YubiKeyType keyType)
+            throws IOException {
+        try {
+            ManagementSession session = new ManagementSession(connection);
+            return session.getDeviceInfo();
+        } catch (CommandException exception) {
+            Logger.d("Unable to get info via Management application, using fallback");
+
+            final Version version =
+                    keyType == YubiKeyType.YKP ?
+                            new Version(4, 0, 0) :
+                            new Version(3, 0, 0);
+
+            Map<Transport, Integer> supportedApps = new EnumMap<>(Transport.class);
+            supportedApps.put(Transport.USB, U2F.bit);
+            if (keyType == YubiKeyType.NEO) {
+                int usbApps = supportedApps.get(Transport.USB);
+                supportedApps.replace(Transport.USB, usbApps | baseNeoApps);
+                supportedApps.put(Transport.NFC, supportedApps.get(Transport.USB));
+            }
+
+            return new DeviceInfo(
+                    new DeviceConfig.Builder().build(), // defaults
+                    null,
+                    version,
+                    FormFactor.USB_A_KEYCHAIN,
+                    supportedApps,
+                    false,
+                    false,
+                    false);
+        }
+    }
+
+    static boolean isPreviewVersion(Version version) {
+        return (version.isAtLeast(5, 0, 0) && version.isLessThan(5, 1, 0))
+                || (version.isAtLeast(5, 2, 0) && version.isLessThan(5, 2, 3))
+                || (version.isAtLeast(5, 5, 0) && version.isLessThan(5, 5, 2));
+    }
+
+    /**
+     * Reads out DeviceInfo from a YubiKey, or attempts to synthesize the data.
+     * <p></p>
+     * Reading DeviceInfo from a ManagementSession is only supported for newer YubiKeys.
+     * This function attempts to read that information, but will fall back to gathering the
+     * data using other mechanisms if needed. It will also make adjustments to the data if
+     * required, for example to "fix" known bad values.
+     * <p></p>
+     * The <code>pid</code> parameter must be provided whenever the YubiKey is connected via USB,
+     *
+     * @param connection {@link SmartCardConnection}, {@link OtpConnection} or
+     * {@link FidoConnection} connection to the YubiKey
+     * @param pid        USB product ID of the YubiKey, can be null if unknown
+     * @throws IOException               in case of connection error
+     * @throws InvalidParameterException in case of <code>pid</code> is null for USB connection
+     * @throws InvalidParameterException in case of connection is not {@link SmartCardConnection}, {@link OtpConnection} or
+     * {@link FidoConnection}
+     */
+    public static DeviceInfo readInfo(YubiKeyConnection connection, @Nullable YubiKeyUsbProductId pid)
+            throws IOException, InvalidParameterException {
```

Comment:
I think this should be IllegalArgumentException rather than InvalidParameterException?